### PR TITLE
feat: seed module that accepts prisma client instance

### PR DESCRIPTION
**Pull request by eemmiillyy**
Part of https://linear.app/prisma-company/issue/CP-1958/update-prismatemplates-seed-function-to-not-rely-on-studio-pcw

This PR replaces the `studio-pcw` package with the `reflector` package. It imports the in memory client factory function from reflector, which allows the usage of a dataproxy connection string instead of a user's connection string.

### CODE REVIEW

File: generator/lib/ArtifactProviders/ExecutableSeed/babel-plugin-transform-template.ts [hunk 1]
```diff
@@ -61,6 +67,7 @@ export type BabelPluginTransformTemplateOptions = {
     content: string
     path: string
     datasourceUrlEnvironmentVariableName: string
+    transformedPrismaSchemaContentEnvironmentVariableName: string
```
Comment (eemmiillyy) [inline generator/lib/ArtifactProviders/ExecutableSeed/babel-plugin-transform-template.ts:70]: Passing the transformed schema contents in like this is not strictly necessary - we could even use the original schema contents now that PCW is gone. However, it's easier to do it like this because the multiline content of is hard to format.

File: generator/lib/ArtifactProviders/ExecutableSeed/babel-plugin-transform-template.ts [hunk 2]
```diff
@@ -106,149 +118,105 @@ export function babelPluginTransformTemplate(
...
+                  Babel.types.identifier('prisma'),
+                  Babel.types.awaitExpression(
+                    Babel.types.newExpression(Babel.types.identifier('PrismaClient'), [])
+                  )
+                ),
+              ])
+            )
+          },
+          AwaitExpression(path) {
+            // Remove prisma.$disconnect(); because it causes an error in the script
```
Comment (eemmiillyy) [inline generator/lib/ArtifactProviders/ExecutableSeed/babel-plugin-transform-template.ts:212]: Prisma is only defined in main and therefore undefined at this part.

File: generator/lib/ArtifactProviders/ExecutableSeed/babel-plugin-transform-template.ts [hunk 3]
```diff
@@ -106,149 +118,105 @@ export function babelPluginTransformTemplate(
...
-                                    Babel.types.stringLiteral('node_modules/.prisma/client/index.js'),
-                                  ]
+                                Babel.types.identifier('contents'),
+                                Babel.types.identifier(
+                                  `process.env.${transformedPrismaSchemaContentEnvironmentVariableName}`
                                 )
                               ),
+                              Babel.types.objectProperty(
+                                Babel.types.identifier('path'),
+                                Babel.types.identifier(`"${schemaPath}"`)
```
Comment (eemmiillyy) [inline generator/lib/ArtifactProviders/ExecutableSeed/babel-plugin-transform-template.ts:190]: The schema path is hardcoded to `/tmp/schema.prisma` however we could improve this by writing it in`tmp/[environmentId]/schema.prisma` . However I think it would have to be passed in through `process.env` from PDP. I'm not sure how to pass a string argument with execa without nesting inside the `env` property

File: package.json [hunk 4]
```diff
@@ -45,8 +45,6 @@
     "tslib": "^2.3.1"
   },
   "devDependencies": {
-    "@babel/core": "7.17.10",
-    "@babel/plugin-transform-typescript": "7.16.8",
```
Comment (jasonkuhrt) [inline package.json:49]: Stop using AST approach and babel.

File: tests/e2e/RentalsPlatform.spec.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import { testTemplate } from './__testers__'
+
+testTemplate({
+  templateName: 'RentalsPlatform',
+  expectedDevOutput: /Expensive reservations:/,
+})
```
Comment (jasonkuhrt) [inline tests/e2e/RentalsPlatform.spec.ts:6]: Adding tests for other templates.

File: tests/e2e/__testers__.ts [hunk 6]
```diff
@@ -75,6 +86,20 @@ export const testTemplate = (params: {
     expect(stripAnsi(initResult.stdout)).toMatch('Generated Prisma Client')
     expect(stripAnsi(initResult.stdout)).toMatch('The seed command has been executed.')
 
+    // TODO only empty template does not have this artifact. Our short circuit above should narrow the type here...
+    if ('prisma/seed.js' in ctx.template.artifacts) {
```
Comment (jasonkuhrt) [inline tests/e2e/__testers__.ts:90]: This narrowing should not be needed, it is a missing feature/bug.

File: tests/e2e/__testers__.ts [hunk 7]
```diff
@@ -75,6 +86,20 @@ export const testTemplate = (params: {
...
+      const prisma = await ctx.getPrisma()
+      try {
+        const seedModuleFile = ctx.template.artifacts['prisma/seed.js']
+        ctx.fs.write('seedModule.js', seedModuleFile.content)
+        const seedModule = require(ctx.fs.path('seedModule.js'))
+        // TODO improve seed scripts to return reports that we can use to capture feedback here not to mention for users generally.
+        await seedModule.run({ prisma })
+      } finally {
+        await prisma.$disconnect()
+      }
```
Comment (jasonkuhrt) [inline tests/e2e/__testers__.ts:100]: We've improved the test suite for all templates to run the derived seed module.

File: generator/lib/ArtifactProviders/ExecutableSeed.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  return [
+    {
+      path: 'prisma/seed.js',
+      content: prismaSeedModule,
+    },
+  ]
+}
+
+const transform = ({ content }: { content: string }) => {
```
Comment (jasonkuhrt) [inline generator/lib/ArtifactProviders/ExecutableSeed.ts:27]: Our transformer now is just regexp and then a pass of SWC to go from TS to JS. No MJS anymore either which simplifies things further.

File: tests/e2e/__testers__.ts [hunk 9]
```diff
@@ -75,6 +86,20 @@ export const testTemplate = (params: {
     expect(stripAnsi(initResult.stdout)).toMatch('Generated Prisma Client')
     expect(stripAnsi(initResult.stdout)).toMatch('The seed command has been executed.')
 
+    // TODO only empty template does not have this artifact. Our short circuit above should narrow the type here...
+    if ('prisma/seed.js' in ctx.template.artifacts) {
+      const prisma = await ctx.getPrisma()
+      try {
+        const seedModuleFile = ctx.template.artifacts['prisma/seed.js']
+        ctx.fs.write('seedModule.js', seedModuleFile.content)
```
Comment (jasonkuhrt) [inline tests/e2e/__testers__.ts:94]: We do not transform the seed module into an eval-able function. Doing so would require a bit more consideration:

1. Eval doesn't work with async? But Node VM module might which anyways we should prefer?
2. Need to turn imports into requires, lots of edge-cases there...

**Review by jasonkuhrt — APPROVED** [review]

**Comment by jasonkuhrt** [comment]
I will see if I can get the Nextjs template e2e tests passing.
